cnd/src/btsieve/ethereum/mod.rs

```diff
@@ -26,321 +30,198 @@ where
+        }
     }
 }
 
-/// Processes block hashes from the `next_hash` receiver, fetches blocks from
-/// the blockchain connector by block hash and enqueues the block on the
-/// `block_queue`. Enqueues the block hash and parent block hash on the
-/// `find_parent_queue`.
-async fn process_blocks_by_hash<C, E>(
-    connector: C,
-    mut block_queue: Sender<Block>,
-    mut find_parent_queue: Sender<(Hash, Hash)>,
-    next_block_channel: (Sender<Hash>, Receiver<Hash>),
-) where
+async fn yield_blocks<C, E>(
+    mut connector: C,
+    co: &Co<Block>,
+    reference_timestamp: Option<u32>,
+) -> anyhow::Result<std::convert::Infallible>
+where
     C: LatestBlock<Block = Option<Block>, Error = E>
         + BlockByHash<Block = Option<Block>, BlockHash = Hash, Error = E>
         + ReceiptByHash<Receipt = Option<TransactionReceipt>, TransactionHash = Hash, Error = E>
         + Clone,
-    E: Debug + Send + 'static,
+    E: std::error::Error + Debug + Send + Sync + 'static,
 {
-    let (mut fetch_block_by_hash_queue, mut next_hash) = next_block_channel;
-    let mut sent_blockhashes: HashSet<H256> = HashSet::new();
+    let mut seen_blockhashes: HashSet<Hash> = HashSet::new();
 
     loop {
-        match next_hash.recv().await {
-            Some(blockhash) => {
-                match connector.block_by_hash(blockhash).compat().await {
-                    Ok(Some(block)) => {
-                        if !sent_blockhashes.contains(&blockhash) {
-                            sent_blockhashes.insert(blockhash);
+        // The duration of this timeout could/should depend on the network
+        tokio::time::delay_for(std::time::Duration::from_secs(1)).await;
 
-                            let _ = join(
-                                block_queue.send(block.clone()),
-                                find_parent_queue.send((blockhash, block.parent_hash)),
-                            )
-                            .await;
-                        }
-                    }
-                    Ok(None) => {
-                        log::warn!("Block with hash {} does not exist", blockhash);
-                    }
-                    Err(e) => {
-                        log::warn!("Could not get block with hash {}: {:?}", blockhash, e);
+        let block = connector
+            .latest_block()
+            .compat()
+            .await?
+            .ok_or_else(|| anyhow::anyhow!("Connector returned nullable latest block"))?;
 
-                        let _ = fetch_block_by_hash_queue.send(blockhash).await;
-                    }
-                };
+        let blockhash = block
+            .hash
+            .ok_or_else(|| anyhow::anyhow!("Connector returned latest block with nullable hash"))?;
+        seen_blockhashes.insert(blockhash);
+
+        if let Some(timestamp) = reference_timestamp {
+            if seen_blockhashes.len() == 1 && block.timestamp > U256::from(timestamp) {
```

Comment:
Can we have a helper function please e.g., `block_is_younger_than_time(block, time)` (note parameter order matches order in function name).  Then future developers do not need to check to think if the `<` and `>` are correct.